pkg/ssh/ssh.go

```diff
@@ -0,0 +1,216 @@
+// Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+package ssh
+
+import (
+	"bytes"
+	"context"
+	"embed"
+	"fmt"
+	"io"
+	"path/filepath"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/ava-labs/avalanche-cli/pkg/constants"
+	"github.com/ava-labs/avalanche-cli/pkg/models"
+	"github.com/ava-labs/avalanche-cli/pkg/ux"
+)
+
+type scriptInputs struct {
+	Log                  string
+	AvalancheGoVersion   string
+	SubnetExportFileName string
+	SubnetName           string
+	GoVersion            string
+	CliBranch            string
+}
+
+//go:embed shell/*.sh
+var script embed.FS
+
+func ScriptLog(line string, nodeID string) string {
+	line = strings.TrimPrefix(line, constants.SSHScriptLogFilter) + " " // add space
+	return fmt.Sprintf("[%s] %s", nodeID, line)
+}
+
+func splitScript(input *bytes.Buffer, separatorPrefix string) ([]*bytes.Buffer, []string, error) {
+	var buffers []*bytes.Buffer
+	separators := []string{""}
+	currentBuffer := &bytes.Buffer{}
+	_, err := currentBuffer.WriteString("#/usr/bin/env bash\nset -euo pipefail\n")
+	if err != nil {
+		return nil, nil, err
+	}
+	for {
+		line, err := input.ReadString('\n')
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, nil, err
+		}
+		if strings.HasPrefix(line, separatorPrefix) {
+			buffers = append(buffers, currentBuffer)
+			currentBuffer = &bytes.Buffer{}
+			_, err := currentBuffer.WriteString("#/usr/bin/env bash\nset -euo pipefail\n")
+			if err != nil {
+				return nil, nil, err
+			}
+			separators = append(separators, line)
+		} else {
+			_, err = currentBuffer.WriteString(line)
+			if err != nil {
+				return nil, nil, err
+			}
+		}
+	}
+	buffers = append(buffers, currentBuffer) // Add the last buffer to the result
+	return buffers, separators, nil
+}
+
+// RunSSHSetupNode runs provided script path over ssh.
+// This script can be template as it will be rendered using scriptInputs vars
+func RunOverSSH(id string, host models.Host, scriptPath string, templateVars scriptInputs) error {
+	templateVars.Log = constants.SSHScriptLogFilter // set log filter
+	shellScript, err := script.ReadFile(scriptPath)
+	if err != nil {
+		return err
+	}
+
+	var script bytes.Buffer
+	t, err := template.New(id).Parse(string(shellScript))
+	if err != nil {
+		return err
+	}
+	err = t.Execute(&script, templateVars)
+	if err != nil {
+		return err
+	}
+	tasks, taskTitle, err := splitScript(&script, fmt.Sprintf("#name:%s", constants.SSHScriptLogFilter))
+	if err != nil {
+		return err
+	}
+	for n, task := range tasks {
+		if taskTitle[n] == "" {
+			continue
+		}
+		ux.Logger.PrintToUser(ScriptLog(strings.TrimPrefix(taskTitle[n], fmt.Sprintf("#name:%s", constants.SSHScriptLogFilter)), host.NodeID))
```

Comment:
there is some \n\n double observable in the output, that should be avoided in this code